duck/ornithology.py

```diff
@@ -9,27 +9,58 @@
 ANY = mock.ANY
 
 
-class Instance(object):
-    """A class that checks equality based on type.
+class Predicate(object):
+    """An object considered to be equal to anything that passes its predicate
+    function.
 
-    Instances of Instance have an .__eq__ method that returns
-    True if and only if the other comparison object is an
-    instance of the provided class (as defined by isinstance).
+    Args:
+        predicate (Callable): The predicate to use to test.
     """
-    def __init__(self, _class):
-        self._class = _class
+    def __init__(self, predicate):
+        self._predicate = predicate
 
     def __eq__(self, other):
-        return isinstance(other, self._class)
+        return self._predicate(other)
 
     def __ne__(self, other):
         return not self == other
 
+
+class Instance(Predicate):
+    """An object considered equal to any instance of the correct type.
+
+    Args:
+        class_ (type): The class to be checked against.
+    """
+    def __init__(self, class_):
+        self._class = class_
+        super(Instance, self).__init__(lambda t: isinstance(t, class_))
+
     def __repr__(self):
         return '<Instance: {0}>'.format(self._class.__name__)
 
 
+class Needle(Predicate):
+    """An object considered equal to any instance in which the compared
+    value is contained within the presented object.
+
+    Args:
+        haystack (object): The object to be searched for the compared value
+    """
+    def __init__(self, haystack):
+        self._object = haystack
+        # todo: we may need to use a helper function to capture an
+        # AttributeError or TypeError here, unless we're okay with the
+        # unhandled exceptions when mismatching types
+        super(Needle, self).__init__(lambda t: haystack.__contains__(t))
```

Comment:
`lambda haystack: needle in haystack` (Use `h` or `hay` if `haystack` is too long, which I suspect it is.)

It is pretty rare that you want to call dunder methods directly. You usually want to _define_ them and then call the Python operators; in this case, that means `needle in haystack` and not `haystack.__contains__(needle)`.